Scripts/Modules/HapticsModule.cs

```diff
@@ -0,0 +1,172 @@
+			// At 320Hz, each sample is 3.125f milliseconds
+			if (Mathf.Approximately(m_MasterIntensity, 0))
+				return;
+
+			m_GeneratedHapticClip.Reset();
+
+			const float kMaxDuration = 0.8f;
+			if (duration > kMaxDuration)
+			{
+				duration = Mathf.Clamp(duration, 0f, kMaxDuration); // Clamp at maxiumum 800ms for sample buffer
+
+				if (!m_SampleLengthWarningShown)
+					Debug.LogWarning("Pulse durations greater than 0.8f are not currently supported");
+
+				m_SampleLengthWarningShown = true;
+			}
+
+			const int kSampleRateConversion = 490; // Samplerate conversion : 44100/90fps = 490
+			const int kIntensityIncreaseMultiplier = 255; // Maximum value of 255 for intensity
+			const float kFadeInProportion = 0.25f;
+			var fadeInSampleCount = duration * kSampleRateConversion * kFadeInProportion;
+			var fadeOutSampleCount = fadeInSampleCount * 2; // FadeOut is less apparent than FadeIn unless FadeOut duration is increased
+			duration *= kSampleRateConversion;
+			var durationFadeOutPosition = duration - fadeOutSampleCount;
+			intensity = Mathf.Clamp(Mathf.Clamp01(intensity) * kIntensityIncreaseMultiplier * m_MasterIntensity, 0, kIntensityIncreaseMultiplier);
+			var hapticClipSample = Convert.ToByte(intensity);
+			for (int i = 1; i < duration; ++i)
+			{
+				float sampleShaped = hapticClipSample;
+				if (fadeIn && i < fadeInSampleCount)
+					sampleShaped = Mathf.Lerp(0, intensity, i / fadeInSampleCount);
+				else if (fadeOut && i > durationFadeOutPosition)
+					sampleShaped = Mathf.Lerp(0, intensity, (duration - i) / fadeOutSampleCount);
+
+				m_GeneratedHapticClip.WriteSample(Convert.ToByte(sampleShaped));
+			}
+
+			const float kMaxSimultaneousClipDuration = 0.25f;
+			var channel = GetTargetChannel(rayOrigin);
+			if (duration > kMaxSimultaneousClipDuration)
+			{
+				// Prevent multiple long clips from playing back simultaneously
+				// If the new clip has a long duration, stop playback of any existing clips in order to prevent haptic feedback noise
+				if (channel != null)
+				{
+					channel.Preempt(m_GeneratedHapticClip);
+				}
+				else
+				{
+					m_RHapticsChannel.Preempt(m_GeneratedHapticClip);
+					m_LHapticsChannel.Preempt(m_GeneratedHapticClip);
+				}
+				
+			}
+			else
+			{
+				// Allow multiple short clips to play simultaneously
+				if (channel != null)
+				{
+					channel.Mix(m_GeneratedHapticClip);
+				}
+				else
+				{
+					m_RHapticsChannel.Mix(m_GeneratedHapticClip);
+					m_LHapticsChannel.Mix(m_GeneratedHapticClip);
+				}
+			}
+#endif
+		}
+
+		public void StopPulses(Transform rayOrigin = null)
+		{
+#if ENABLE_OVR_INPUT
+			// Stop haptic feedback on only one device, if specific via the function parameter
+			// Otherwise stop all haptic feedback on all devices
+			if (rayOrigin == null)
+			{
+				m_RHapticsChannel.Clear();
+				m_LHapticsChannel.Clear();
+			}
+			else
+			{
+				var channel = GetTargetChannel(rayOrigin);
+				if (channel != null)
```

Comment:
I'll add a warning.  The implementation is intentional, so as to handle not clearing any existing valid channel playback if an invalid rayorigin is passed in.  This will allow haptic playback to continue in such a case (especially for longer haptic effects).